// @imagekit/react/dist/imagekitio-react.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var javascript = require('@imagekit/javascript');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

const ImageKitContext = React.createContext({});
/**
 * You can use the `ImageKitProvider` component to set the default values for all nested Image and Video components provided by the SDK.
 *
 * It accepts the following props:
 * - `urlEndpoint`: The URL endpoint for your ImageKit account. You can get this from the [ImageKit dashboard](https://imagekit.io/dashboard/url-endpoints).
 * - `transformationPosition`: The position of the transformation string in the URL. It can be either `path` or `query`. By default, it is set to `query`.
 *
 * @example
 * ```jsx
 * import { ImageKitProvider, Image } from "@imagekit/react";
 * <ImageKitProvider urlEndpoint="https://ik.imagekit.io/your_imagekit_id">
 *   <Image src="/default-image.jpg" />
 * </ImageKitProvider>
 * ```
 */
const ImageKitProvider = props => {
  const {
    urlEndpoint,
    transformationPosition
  } = props;
  return React__default["default"].createElement(ImageKitContext.Provider, {
    value: {
      urlEndpoint,
      transformationPosition
    }
  }, props.children);
};

function getInt(x) {
  if (typeof x === 'undefined') {
    return NaN;
  }
  if (typeof x === 'number') {
    return Number.isFinite(x) ? x : NaN;
  }
  if (typeof x === 'string' && /^[0-9]+$/.test(x)) {
    return parseInt(x, 10);
  }
  return NaN;
}
/**
 * A React wrapper around the native `<img>` element that adds ImageKit's
 * optimization and transformation capabilities.
 *
 * All native `<img>` props are supported, in addition to the ImageKit‑specific props.
 *
 * @example
 * ```jsx
 * import { Image } from "@imagekit/react";
 *
 * <Image
 *   urlEndpoint="https://ik.imagekit.io/your_imagekit_id" // Can also be set on a parent <ImageKitProvider />
 *   src="/default-image.jpg"                               // Path in your ImageKit account
 *   alt="Default image"
 *   width={500}
 *   height={500}
 *   transformation={[{ width: 500, height: 500 }]}         // Optional ImageKit transformations
 * />
 * ```
 */
const Image = props => {
  const contextValues = React.useContext(ImageKitContext);
  // It's important to extract the ImageKit‑specific props so we can spread the
  // remaining props directly onto the `<img>` element.
  const _a = Object.assign(Object.assign({}, contextValues), props // Override with props
    ),
    {
      transformation = [],
      src = "",
      loading = "lazy",
      queryParameters,
      urlEndpoint,
      transformationPosition,
      sizes,
      responsive = true,
      deviceBreakpoints,
      imageBreakpoints
    } = _a,
    nonIKParams = __rest(_a, ["transformation", "src", "loading", "queryParameters", "urlEndpoint", "transformationPosition", "sizes", "responsive", "deviceBreakpoints", "imageBreakpoints"]);
  // Fail fast in development if the mandatory urlEndpoint is missing.
  if (!urlEndpoint || urlEndpoint.trim() === "") {
    if (process.env.NODE_ENV !== "production") {
      console.error("urlEndpoint is neither provided in this component nor in any parent ImageKitProvider.");
    }
    return null;
  }
  const widthInt = getInt(nonIKParams.width);
  if (!responsive) {
    const nonResponsiveSrc = javascript.buildSrc({
      src,
      transformation,
      queryParameters,
      urlEndpoint,
      transformationPosition
    });
    return React__default["default"].createElement("img", Object.assign({}, nonIKParams, {
      loading: loading,
      src: nonResponsiveSrc
    }));
  }
  const {
    src: newSrc,
    srcSet
  } = javascript.getResponsiveImageAttributes({
    src,
    transformation,
    width: isNaN(widthInt) ? undefined : widthInt,
    sizes,
    queryParameters,
    urlEndpoint,
    transformationPosition,
    deviceBreakpoints,
    imageBreakpoints
  });
  return React__default["default"].createElement("img", Object.assign({}, nonIKParams, {
    loading: loading,
    sizes: sizes,
    srcSet: srcSet,
    src: newSrc
  }));
};

/**
 * The Video component is a wrapper around the HTML video element. It supports all the features of the HTML video element, along with additional features provided by ImageKit.
 *
 * @example
 * ```jsx
 * import { Video } from "@imagekit/react";
 * <Video
 *  urlEndpoint="https://ik.imagekit.io/your_imagekit_id" // You can also set this in a parent ImageKitProvider component
 *  src="/default-video.mp4" // The path to the video in your ImageKit account
 *  controls
 *  width={500}
 *  height={500}
 *  transformation={[{ width: 500, height: 500 }]} // Add ImageKit transformations
 * />
 * ```
 */
const Video = props => {
  const contextValues = React.useContext(ImageKitContext);
  // Extract ImageKit‑specific props so we can spread the rest directly onto <video>.
  const _a = Object.assign(Object.assign({}, contextValues), props // Override with props
    ),
    {
      transformation = [],
      src = "",
      queryParameters,
      urlEndpoint,
      transformationPosition
    } = _a,
    nonIKParams = __rest(_a, ["transformation", "src", "queryParameters", "urlEndpoint", "transformationPosition"]);
  // Fail fast in development if the mandatory urlEndpoint is missing.
  if (!urlEndpoint || urlEndpoint.trim() === "") {
    if (process.env.NODE_ENV !== "production") {
      console.error("urlEndpoint is neither provided in this component nor in any parent ImageKitProvider.");
    }
    return null;
  }
  const finalSrc = javascript.buildSrc({
    urlEndpoint,
    src,
    transformation: [...transformation],
    // Do not mutate original transformation array from the props
    queryParameters,
    transformationPosition
  });
  return React__default["default"].createElement("video", Object.assign({}, nonIKParams, {
    src: finalSrc
  }));
};

Object.defineProperty(exports, 'ImageKitAbortError', {
    enumerable: true,
    get: function () { return javascript.ImageKitAbortError; }
});
Object.defineProperty(exports, 'ImageKitInvalidRequestError', {
    enumerable: true,
    get: function () { return javascript.ImageKitInvalidRequestError; }
});
Object.defineProperty(exports, 'ImageKitServerError', {
    enumerable: true,
    get: function () { return javascript.ImageKitServerError; }
});
Object.defineProperty(exports, 'ImageKitUploadNetworkError', {
    enumerable: true,
    get: function () { return javascript.ImageKitUploadNetworkError; }
});
Object.defineProperty(exports, 'buildSrc', {
    enumerable: true,
    get: function () { return javascript.buildSrc; }
});
Object.defineProperty(exports, 'buildTransformationString', {
    enumerable: true,
    get: function () { return javascript.buildTransformationString; }
});
Object.defineProperty(exports, 'getResponsiveImageAttributes', {
    enumerable: true,
    get: function () { return javascript.getResponsiveImageAttributes; }
});
Object.defineProperty(exports, 'upload', {
    enumerable: true,
    get: function () { return javascript.upload; }
});
exports.Image = Image;
exports.ImageKitContext = ImageKitContext;
exports.ImageKitProvider = ImageKitProvider;
exports.Video = Video;
